lib/APPTurtle/src/CustomRosTransport.cpp

```diff
@@ -0,0 +1,211 @@
+
+    if (nullptr == transport)
+    {
+        LOG_ERROR("Transport is nullptr.");
+    }
+    else
+    {
+        const micro_ros_agent_locator* locator = reinterpret_cast<micro_ros_agent_locator*>(transport->args);
+        const int                      UDP_OK  = 1;
+
+        if (UDP_OK != udpClient.begin(locator->port))
+        {
+            LOG_ERROR("UDP begin error");
+        }
+        else
+        {
+            isOpen = true;
+        }
+    }
+
+    return isOpen;
+}
+
+bool custom_transport_close(uxrCustomTransport* transport)
+{
+    UTIL_NOT_USED(transport);
+
+    udpClient.stop();
+    return true;
+}
+
+size_t custom_transport_write(uxrCustomTransport* transport, const uint8_t* buffer, size_t size, uint8_t* errorCode)
+{
+    size_t sent = 0;
+
+    if ((nullptr == transport) || (nullptr == buffer) || (0 == size) || (nullptr == errorCode))
+    {
+        LOG_ERROR("One or more parameters are invalid.");
+    }
+    else
+    {
+        const micro_ros_agent_locator* locator = reinterpret_cast<micro_ros_agent_locator*>(transport->args);
+        const int                      UDP_OK  = 1;
+        int                            ret     = UDP_OK;
+
+        ret = udpClient.beginPacket(locator->address, locator->port);
+
+        if (UDP_OK != ret)
+        {
+            LOG_ERROR("UDP beginPacket error");
+        }
+        else
+        {
+            size_t bytesToWrite = udpClient.write(buffer, size);
+
+            if (bytesToWrite != size)
+            {
+                LOG_ERROR("UDP write error");
+            }
+            else
+            {
+                ret = udpClient.endPacket();
+
+                if (UDP_OK != ret)
+                {
+                    LOG_ERROR("UDP endPacket error");
+                }
+                else
+                {
+                    sent = bytesToWrite;
+                }
+            }
+        }
+
+        if (UDP_OK != ret)
+        {
+            int writeErrorCode = udpClient.getWriteError();
+            *errorCode         = writeErrorCode;
+            LOG_ERROR("UDP error: %d", writeErrorCode);
+        }
+
+        udpClient.flush();
+    }
+
+    return sent;
+}
+
+size_t custom_transport_read(uxrCustomTransport* transport, uint8_t* buffer, size_t size, int timeout,
+                             uint8_t* errorCode)
+{
+    UTIL_NOT_USED(transport);
+    size_t readBytes = 0;
+
+    if ((nullptr == buffer) || (0 == size) || (0 == timeout) || (nullptr == errorCode))
+    {
+        LOG_ERROR("One or more parameters are invalid.");
+    }
+    else
+    {
+        SimpleTimer readTimer;
+        readTimer.start(timeout);
+
+        while (false == readTimer.isTimeout())
+        {
+            if (0 < udpClient.parsePacket())
+            {
+                break;
+            }
+        }
+
+        if (0 == udpClient.available())
+        {
+            *errorCode = 1;
+        }
+        else
+        {
+            readBytes = udpClient.read(buffer, size);
```

Comment:
what happen if we received more bytes then read out here ? I assume we block next loop until timeout also data is still available.